bokehjs/src/lib/models/tools/toolbar.ts

```diff
@@ -122,22 +122,60 @@ export class ToolbarView extends UIElementView {
     super.remove()
   }
 
+  readonly _our_proxies: ToolProxy<Tool>[] = []
+
   protected async _build_tool_button_views(): Promise<void> {
+    for (const proxy of this._our_proxies) {
+      proxy.destroy()
+    }
+    clear(this._our_proxies)
+
     this._tool_buttons = (() => {
       const {buttons} = this.model
       if (buttons == "auto") {
-        const groups = [
+        const tool_bars: ToolLike<Tool>[][] = [
           ...values(this.model.gestures).map((gesture) => gesture.tools),
           this.model.actions,
           this.model.inspectors,
           this.model.auxiliaries,
         ]
-        const buttons = groups.map((group) => {
-          return group
+
+        const {group, group_types} = this.model
+        const button_bars = tool_bars.map((bar) => {
+          const tools = (() => {
+            if (group) {
+              const grouped: Map<string, ToolLike<Tool>[]> = new Map()
+              for (const [tool, i] of enumerate(bar)) {
+                const group_type = group_types.find((type) => Tool.is_alias_of(tool, type))
+                if (group_type != null) {
+                  const group = grouped.get(tool.type)
+                  if (group != null) {
+                    group.push(tool)
+                  } else {
+                    grouped.set(tool.type, [tool])
+                  }
+                } else {
+                  grouped.set(`${i}`, [tool])
+                }
+              }
+              return [...grouped.values()].map((group) => {
```

Comment:
Can `Array.from` be used here?

Reply:
You mean `Array.from(iterator, fn)`? Sure. Is it going to be faster? Maybe, though you never know how slow certain native APIs can be.